Annotate errors in apimachinery e2e tests

audreylim opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->
Related to https://github.com/kubernetes/kubernetes/issues/34059

**What this PR does / why we need it**:
This PR makes errors more helpful in the apimachinery e2e tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
My first contribution to Kubernetes

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

audreylim committed:
Annotate errors in apimachinery e2e tests

wgliang commented:
/ok-to-test

audreylim commented:
/retest

ixdy commented:
/assign
/retest

nikhita commented:
/assign @spiffxp

spiffxp commented on test/e2e/apimachinery/etcd_failure.go:
```diff
@@ -130,7 +131,7 @@ func checkExistingRCRecovers(f *framework.Framework) {
 		options := metav1.ListOptions{LabelSelector: rcSelector.String()}
 		pods, err := podClient.List(options)
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to list pods in namespace: %s, that match label selector: %s", f.Namespace.Name, rcSelector.String())
```
Overall this PR LGTM, but I'm curious if there is a way we could get the namespace name from `podClient` here? Something about having to refer back N>10 lines to see why/how to craft this message seems less than ideal

spiffxp commented on test/e2e/apimachinery/initializers.go:
```diff
@@ -72,34 +73,34 @@ var _ = SIGDescribe("Initializers [Feature:Initializers]", func() {
 		// pod should now start running
 		err = framework.WaitForPodRunningInNamespace(c, pod)
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "timeout while waiting for pod %s to run in namespace: %s, or pod is in a Failed state", pod.Name, pod.Namespace)
```
maybe `"error while waiting for pod %s to go to Running phase in namespace %s"`

spiffxp commented on test/e2e/apimachinery/initializers.go:
```diff
@@ -72,34 +73,34 @@ var _ = SIGDescribe("Initializers [Feature:Initializers]", func() {
 		// verify that we cannot start the pod initializing again
 		pod, err = c.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to get pod: %s", podName)
```
missing namespace

spiffxp commented on test/e2e/apimachinery/initializers.go:
```diff
@@ -145,8 +147,9 @@ var _ = SIGDescribe("Initializers [Feature:Initializers]", func() {
-			Expect(err).NotTo(HaveOccurred())
+			pod := newInitPod(podName)
+			_, err := c.CoreV1().Pods(ns).Create(pod)
+			Expect(err).NotTo(HaveOccurred(), "failed to create pod: %+v", pod)
```
missing namespace, and consider using the pod's name with a `%s` instead

spiffxp commented on test/e2e/apimachinery/initializers.go:
```diff
@@ -72,34 +73,34 @@ var _ = SIGDescribe("Initializers [Feature:Initializers]", func() {
 		// verify the list call filters out uninitialized pods
 		pods, err := c.CoreV1().Pods(ns).List(metav1.ListOptions{IncludeUninitialized: true})
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to list pods in namespace: %s", ns)
```
I saw earlier messages included the list options, maybe here as well?

spiffxp reviewed (changes requested):
Thanks so much for PR'ing this, this is a lot of lines touched!

I reviewed about halfway before stopping, so keep in mind there are more examples of the following than I explicitly called out:
- I think we want to be consistent about logging the namespace being used for namespace-scoped objects
- I think we generally don't want to dump an entire pod spec with `+%v`, but there were some cases that may make sense (ie: error updating pod "foo" to <pod spec>)
- I'm not sure it's worth calling out "or pods went to Failed state" if you're err'ing on WaitForPod, but I could be convinced either way

audreylim commented on test/e2e/apimachinery/garbage_collector.go:
```diff
@@ -818,31 +818,34 @@ var _ = SIGDescribe("Garbage collector", func() {
 		deleteOptions.Preconditions = metav1.NewUIDPreconditions(string(pod1.UID))
 		err = podClient.Delete(pod1.ObjectMeta.Name, deleteOptions)
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to delete pod: %s", pod1.ObjectMeta.Name)
```
I referenced the previous line which used `pod1.ObjectMeta.Name` but I think both refer to the same since `ObjectMeta.Name` is a promoted field (https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3116). I've changed it to `pod1.Name` here and left the other references to `pod1.ObjectMeta.Name` alone, maybe the author wanted more clarity.

spiffxp commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -672,15 +672,15 @@ func testWebhook(f *framework.Framework) {
 	}
 	_, err = updateConfigMap(client, f.Namespace.Name, allowedConfigMapName, toNonCompliantFn)
-	Expect(err).NotTo(BeNil())
+	Expect(err).To(HaveOccurred(), "update (PUT) admitted configmap %s to a non-compliant one should be rejected by webhook", allowedConfigMapName)
```
missing namespace

spiffxp commented on test/e2e/apimachinery/watch.go:
```diff
@@ -261,23 +263,23 @@ var _ = SIGDescribe("Watchers", func() {
 		By("creating a new configmap")
 		testConfigMap, err = c.CoreV1().ConfigMaps(ns).Create(testConfigMap)
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to create configmap %s", configMapName, configMapName)
```
missing namespace

audreylim commented:
Thanks for your feedback! I've:
- added namespaces where its scoped
- replaced spec dumps with object names
- addressed the other comments

Notes:
- I added `or pods went to Failed state` on WaitForPod to give possible reasons on failure (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1422), but I changed them to the more generic and also helpful "error while waiting for pod %s to go to Running phase in namespace %s". 
- I can squash the commits if needed

spiffxp commented on test/e2e/apimachinery/watch.go:
```diff
@@ -261,23 +263,23 @@ var _ = SIGDescribe("Watchers", func() {
 		By("creating a watch on configmaps with a certain label")
 		testWatch, err := watchConfigMaps(f, "", toBeChangedLabelValue)
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to create configmap with label: %s", toBeChangedLabelValue)
```
failed to create a watch on configmaps with label

spiffxp commented on test/e2e/apimachinery/watch.go:
```diff
@@ -197,17 +198,17 @@ var _ = SIGDescribe("Watchers", func() {
 		By("creating a watch on configmaps")
 		testWatchBroken, err := watchConfigMaps(f, "", watchRestartedLabelValue)
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), "failed to create a watch on configmaps: %s", watchRestartedLabelValue)
```
configmaps with with label

spiffxp reviewed (commented):
few tiny nits to fix, otherwise lgtm

and yeah feel free to squash

audreylim commented:
Fixed

audreylim commented:
oops, made a slight change here: https://github.com/kubernetes/kubernetes/pull/69583/files#diff-b3ef5995f1063613af33e678183dc03dR57

removed `pod := newTablePod(podName)` since pod spec wasn't used in error message anymore

spiffxp commented:
/lgtm
